components/zcash_address/src/kind/unified.rs

```diff
@@ -24,32 +28,156 @@ pub(crate) const REGTEST: &str = "uregtest";
 
 const PADDING_LEN: usize = 16;
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum Typecode {
+    P2pkh,
+    P2sh,
+    Sapling,
+    Orchard,
+    Unknown(u8),
+}
+
+impl Ord for Typecode {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        match (self, other) {
+            // Trivial equality checks.
+            (Self::Orchard, Self::Orchard)
+            | (Self::Sapling, Self::Sapling)
+            | (Self::P2sh, Self::P2sh)
+            | (Self::P2pkh, Self::P2pkh) => cmp::Ordering::Equal,
+
+            // We don't know for certain the preference order of unknown receivers, but it
+            // is likely that the higher typecode has higher preference. The exact order
+            // doesn't really matter, as unknown receivers have lower preference than
+            // known receivers.
+            (Self::Unknown(a), Self::Unknown(b)) => b.cmp(a),
+
+            // For the remaining cases, we rely on `match` always choosing the first arm
+            // with a matching pattern. Patterns below are listed in priority order:
+            (Self::Orchard, _) => cmp::Ordering::Less,
+            (_, Self::Orchard) => cmp::Ordering::Greater,
+
+            (Self::Sapling, _) => cmp::Ordering::Less,
+            (_, Self::Sapling) => cmp::Ordering::Greater,
+
+            (Self::P2sh, _) => cmp::Ordering::Less,
+            (_, Self::P2sh) => cmp::Ordering::Greater,
+
+            (Self::P2pkh, _) => cmp::Ordering::Less,
+            (_, Self::P2pkh) => cmp::Ordering::Greater,
+        }
+    }
+}
+
+impl PartialOrd for Typecode {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl From<u8> for Typecode {
+    fn from(typecode: u8) -> Self {
+        match typecode {
+            0x00 => Typecode::P2pkh,
+            0x01 => Typecode::P2sh,
+            0x02 => Typecode::Sapling,
+            0x03 => Typecode::Orchard,
+            _ => Typecode::Unknown(typecode),
+        }
+    }
+}
+
+impl From<Typecode> for u8 {
+    fn from(t: Typecode) -> Self {
+        match t {
+            Typecode::P2pkh => 0x00,
+            Typecode::P2sh => 0x01,
+            Typecode::Sapling => 0x02,
+            Typecode::Orchard => 0x03,
+            Typecode::Unknown(typecode) => typecode,
+        }
+    }
+}
+
+impl Typecode {
+    fn is_shielded(&self) -> bool {
+        match self {
+            Typecode::P2pkh | Typecode::P2sh => false,
+            // Assume that unknown typecodes are shielded, because they might be.
+            _ => true,
```

Comment:
It'd be better to explicitly list the remaining typecodes here, just for future-proofness. This reminds me, how will TZE interact with the addressing scheme?

Reply:
Note that in [zcash/zips#536](https://github.com/zcash/zips/pull/536/files#diff-8ff1060cb8fa47eb429fb6c849f1364e28fd4b42868ae8f56dd20ea8b9028b2cR368-R372), I clarified the requirement that uses this as follows:
> * A Unified Address or Unified Viewing Key MUST NOT contain only
>   transparent P2SH or P2PKH addresses (Typecodes :math:`\mathtt{0x00}`
>   and :math:`\mathtt{0x01}`). The rationale is that the existing
>   P2SH and P2PKH transparent-only address formats suffice for this
>   purpose and are already supported by the existing ecosystem.

So it is explicitly only P2SH and P2PKH addresses that are treated as transparent here (which fits with the rationale).

Reply:
@nuttycom we cannot list the remaining typecodes here, because they include `Typecode::Unknown` which is explicitly unbounded. As for TZE: we would need to decide at the time whether the property we want is "UA must contain a shielded component" (which would preclude TZE-only addrs) or "UA must not contain only a t-address-equivalent component" (which would enable the creation of non-shielded UAs). I suspect the latter is what we will want, treating UAs as the adapter mechanism for plugging in new receivers rather than needing to define a new address encoding every time.

Reply:
The presence of Typecode::Unknown makes me tempted to say that this function should actually return an `Option<Bool>` because otherwise callers can't use it in isolation of other checks against the typecode. This is probably fine; it just means that this won't be flagged by the compiler as a call site that needs inspection if the set of receivers changes. I'm just lazy, I want the compiler to do that work for me, rather than having to remember. :)